yaksh/views.py

```diff
@@ -612,37 +611,71 @@ def check(request, q_id, attempt_num=None, questionpaper_id=None):
             if current_question.type == 'code' or \
             current_question.type == 'upload' else None
         result = paper.validate_answer(
-            user_answer, current_question, json_data
+            user_answer, current_question, json_data, uid
         )
-        if result.get('success'):
-            new_answer.marks = (current_question.points * result['weight'] /
-                current_question.get_maximum_test_case_weight()) \
-                if current_question.partial_grading and \
-                current_question.type == 'code' or current_question.type == 'upload' \
-                else current_question.points
-            new_answer.correct = result.get('success')
-            error_message = None
-            new_answer.error = json.dumps(result.get('error'))
-            next_question = paper.add_completed_question(current_question.id)
+        if current_question.type in ['code', 'upload']:
+            if paper.time_left() <= 0:
+                url = 'http://localhost:%s' % SERVER_POOL_PORT
+                result = get_result(url, uid, block=True)
+                result = json.loads(result.get('result'))
+                next_question, error_message, paper = _update_paper(request, uid,
+                        result)
+                return show_question(request, next_question, paper, error_message)
+            else:
+                return JsonResponse(result)
         else:
-            new_answer.marks = (current_question.points * result['weight'] /
-                current_question.get_maximum_test_case_weight()) \
-                if current_question.partial_grading and \
-                current_question.type == 'code' or current_question.type == 'upload' \
-                else 0
-            error_message = result.get('error') if current_question.type == 'code' \
-                or current_question.type == 'upload' else None
-            new_answer.error = json.dumps(result.get('error'))
-            next_question = current_question if current_question.type == 'code' \
-                or current_question.type == 'upload' \
-                else paper.add_completed_question(current_question.id)
-        new_answer.save()
-        paper.update_marks('inprogress')
-        paper.set_end_time(timezone.now())
-        return show_question(request, next_question, paper, error_message)
+            next_question, error_message, paper = _update_paper(request, uid, result)
+            return show_question(request, next_question, paper, error_message)
     else:
         return show_question(request, current_question, paper)
 
+
+@csrf_exempt
+def get_results(request, uid):
+    result = {}
+    url = 'http://localhost:%s' % SERVER_POOL_PORT
+    result_state = get_result(url, uid)
```

Comment:
What is this get_result?  This is bound to cause confusion.  Maybe `import get_result as get_result_from_code_server`